More Flexible Configs

bbakernoaa opened the pull request:
## Description
 <!-- Provide a brief description of what this PR does, and why it is needed. -->
This pull request introduces a new configuration file for the HEMCO (Harmonized Emissions Component) system, which defines emissions inventories, file paths, and various settings for atmospheric modeling. The changes primarily focus on adding a comprehensive configuration file (`HEMCO_Config.rc`) with detailed sections for settings, extension switches, base emissions, scale factors, and masks.

### Addition of a new HEMCO configuration file:

* **File introduction and structure**:
  - Added a new file `config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc` to define HEMCO configurations, including emissions inventories and file paths. The file is well-documented with comments and references to the HEMCO User's Guide and revision history.

* **Settings section**:
  - Defined global settings such as input/output directories, grid and spec files, diagnostic options, and meteorological resolution (`MET` and `RES`).

* **Extension switches**:
  - Configured various emission extensions (e.g., `HTAPv3`, `NEMO_NEI2019_AR`, `CEDSv2`) with on/off switches and additional parameters. Some extensions like `MEGAN` include specific scaling and inhibition factors.

* **Base emissions**:
  - Included references to external configuration files for multiple emission inventories

 Fixes # (issue)

 ## Type of change
 <!-- Please delete options that are not relevant. -->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - [ ] Documentation update
 - [ ] Python files
 - [ ] Module or Library
 - [ ] NUOPC interface issue

 ## How Has This Been Tested? and what Platform
 <!-- Please describe the tests that you ran to verify your changes. Provide instructions to reproduce and details of your test configuration. -->

 ### Platform Tested

 - [ ] HERA
 - [ ] WCOSS
 - [ ] GAEA
 - [ ] DERECHO
 - [ ] ORION
 - [ ] HERCULES
 - [ ] HOPPER
 - [ ] ASMD

 ## Checklist:
 - [ ] My code follows the style guidelines of this project
 - [ ] I have performed a self-review of my own code
 - [ ] I have commented my code, particularly in hard-to-understand areas
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [ ] New and existing unit tests pass locally with my changes
 - [ ] Any dependent changes have been merged and published in downstream modules

 ## Additional Context
 <!-- Add any other context or screenshots about the pull request here. -->

bbakernoaa committed:
Add HEMCO configuration files for NEMO_NEI2019_AR and NOAA_GMD_CH4 emissions

- Created HEMCO_Config.rc.NEMO_NEI2019_AR to configure emissions from the NEI 2019 dataset for the NEMO model.
- Added detailed settings, extension switches, and base emissions for various pollutants including CO, NOx, SO2, and VOCs.
- Introduced HEMCO_Config.rc.NOAA_GMD_CH4 to configure methane emissions from NOAA GMD monthly mean surface data.
- Included necessary settings and base emissions for CH4 from 1979 to 2020.

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc:
This should be cb6r**5**_ae7.

bbakernoaa replied:
renamed

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc:
```diff
@@ -0,0 +1,176 @@
+# %%% This is the default global inventory. You may select either CEDS,
+# EDGAR, HTAP or CMIP6_SFC_LAND_ANTHRO for the global base emissions %%%
+#==============================================================================
+>>>include $ROOT/Configs/HEMCO_Config.rc.CEDSv2
```
I like this idea of having different files point to the specific configuration, i.e., more flexible.  However, technically this CEDSv2 config contains speciation directly related to CB6 chem mech so the file names should likely be more granular, e.g., $ROOT/Configs/HEMCO_Config.rc.CEDSv2.**CB6** or something like this here and below for other inventory configs.

bbakernoaa replied:
I've reordered some things so that it's more clear.  All configurations are under the mechanism name folder

so `config/cmaq/cb6r5_ae7/*.rc*

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc:
```diff
@@ -0,0 +1,176 @@
+# %%% This is the default global inventory. You may select either CEDS,
+# EDGAR, HTAP or CMIP6_SFC_LAND_ANTHRO for the global base emissions %%%
+#==============================================================================
+>>>include $ROOT/Configs/HEMCO_Config.rc.CEDSv3
```
Maybe you can explain a bit more here, but is the intention for user to comment out either CEDSv2 or CEDv3 depending on desired application?  Would need to be careful to avoid adding and double counting emissions.

bbakernoaa replied:
The intention is to have available different configurations.  Users can then easily modify to include different combinations more easily. There will of course be a "recommended" configuration which will be our target.

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc:
```diff
@@ -0,0 +1,176 @@
+>>>include $ROOT/Configs/HEMCO_Config.rc.NEMO_NEI2019_AR
+
+#==============================================================================
+# -- HTAPv2 ---
```
Same comment as above, why is both HTAPv2 and HTAPv3 included here, wouldn't we want to only include v3 for this config?

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc:
```diff
@@ -0,0 +1,176 @@
+#==============================================================================
+>>>include $ROOT/Configs/HEMCO_Config.rc.MEGAN
+
+
```
Where is [OMI-HTAP SO2 2019](https://github.com/noaa-oar-arl/NEXUS/blob/develop/config/cmaq_gfs_megan_nei2019_globtempo/NEXUS_Config.rc#L565)?  We have been using this as we use it for all land SO2 emissions outside of CONUS.

bbakernoaa replied:
Haven't created it yet but we should also think about this as there are now updated datasets past 2019.

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc.CEDSv2:
Need [OMI_HTAP_MASK](https://github.com/noaa-oar-arl/NEXUS/blob/develop/config/cmaq_gfs_megan_nei2019_globtempo/NEXUS_Config.rc#L1587) to use with CEDS (if adding OMI-HTAP SO2 2019 config).

Not sure if you are abandoning using OMI-HTAP, but the ScalID of 1008 is still included for CEDS SO2 above.

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc.CEDSv2:
```diff
@@ -0,0 +1,629 @@
+# ExtNr ExtName           on/off  Species
+0       Base              : on    *
+    --> HEMCO_RESTART     :       true
+    --> CEDS              :  on   true
```
```suggestion
    --> CEDSv2              :  on   true
```

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc.CEDSv2:
```diff
@@ -0,0 +1,629 @@
+###############################################################################
+### BEGIN SECTION SETTINGS
```
Are section settings needed to be repeated for each inventory/mech config here and in other files below?  I think only the main mech HEMCO_config file would need these settings in the final config.

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc.CEDSv2:
```diff
@@ -0,0 +1,629 @@
+###     Obtain NAP emissions using scale factor from Hays 2002 ES&T
+###     (0.0253 g NAP/kg DM) and Andreae and Merlet 2001 GBC (92 g CO/kg DM)
+###############################################################################
+# ExtNr ExtName           on/off  Species
```
Same comment for Extensions switches.  Are they needed to be repeated in each inventory config file here and below?

bbakernoaa replied:
I believe every configuration file needs these sections for the config parser to work properly.

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc.NEMO_NEI2019_AR:
Missing [NEI and CONUS masks](https://github.com/noaa-oar-arl/NEXUS/blob/develop/config/cmaq_gfs_megan_nei2019_globtempo/NEXUS_Config.rc#L1590-L1593).

drnimbusrain commented on config/model/cmaq/mech/cb6r3_ae7/HEMCO_Config.rc.MEGAN:
```diff
@@ -0,0 +1,333 @@
+###############################################################################
+
+# ScalID Name sourceFile sourceVar sourceTime C/R/E SrcDim SrcUnit Oper
+
```
Missing the u[nits carbon to species conversions](https://github.com/noaa-oar-arl/NEXUS/blob/develop/config/cmaq_gfs_megan_nei2019_globtempo/NEXUS_Config.rc#L1252-L1256) scale factors.

drnimbusrain reviewed (commented):
Hi Barry, interesting development on more flexible config approach!  I have some comments and questions on how its being included.

bbakernoaa committed:
Add CEDS preprocessing script and batch processing shell script

- Implemented `ceds_preprocess.py` to convert sector-based emissions in CEDS files into separate variables for each sector.
- Added functionality to preserve global attributes and handle sector mappings.
- Created `process_ceds_files.sh` for batch processing of CEDS emission files, supporting both interactive and SLURM batch modes.
- Included error handling and logging for better traceability during processing.

bbakernoaa committed:
reorder directories

bbakernoaa committed:
Add HEMCO configuration files for NEMO NEI2019 and NOAA GMD CH4 emissions

- Created HEMCO_Config.rc.NEMO_NEI2019_AR for NEI 2019 emissions with detailed settings and base emissions for various pollutants.
- Added HEMCO_Config.rc.NOAA_GMD_CH4 for NOAA GMD CH4 emissions, including monthly mean surface CH4 data.
- Introduced HEMCO_sa_Grid.rc to define emission grid specifications and vertical coordinate parameters.

bbakernoaa committed:
Update HEMCO configuration to enable CEDSv2 extension
